File: non_profit/non_profit/overrides/server/employee.py
import frappe
from frappe import _
from frappe.model.document import Document


@frappe.whitelist()
def after_insert(doc: Document, method: str) -> None:
    """
    Handle after_insert event for Employee document
    Creates user account and updates employee fields from Job Applicant or Volunteer Signup
    """
    try:
        needs_save = False

        if doc.personal_email and not doc.user_id:
            user_id = create_user_for_employee(doc)
            if user_id:
                doc.user_id = user_id
                needs_save = True

        if doc.job_applicant:
            if update_fields_from_applicant(doc):
                needs_save = True

            if doc.is_volunteer:
                ensure_job_offer_for_volunteer(doc.job_applicant)

        if needs_save:
            doc.save(ignore_permissions=True)
            frappe.db.commit()

    except Exception as e:
        frappe.log_error(f"Error in Employee after_insert for {doc.name}", f"{str(e)}")
        frappe.throw(_("Error processing employee creation: {0}").format(str(e)))


def ensure_job_offer_for_volunteer(job_applicant: str) -> None:
    """
    Ensure a Job Offer exists for the given Volunteer.
    If not found, auto-create from linked Job Applicant.
    """
    if not job_applicant:
        return

    existing_offer = frappe.db.exists(
        "Job Offer",
        {"job_applicant": job_applicant},
    )
    if existing_offer:
        return

    applicant = frappe.get_doc("Job Applicant", job_applicant)

    designation = applicant.designation
    if not designation:
        if not frappe.db.exists("Designation", {"designation_name": "Volunteer"}):
            designation_doc = frappe.get_doc(
                {
                    "doctype": "Designation",
                    "designation_name": "Volunteer",
                }
            )
            designation_doc.insert(ignore_permissions=True)
        designation = "Volunteer"

    job_offer = frappe.new_doc("Job Offer")
    job_offer.job_applicant = applicant.name
    job_offer.company = applicant.company
    job_offer.designation = designation

    job_offer.status = "Accepted"
    job_offer.offer_date = frappe.utils.today()

    job_offer.insert(ignore_permissions=True)
    job_offer.submit()
    frappe.db.commit()


def create_user_for_employee(doc: Document) -> str:
    """
    Create a user account for the employee
    Returns the user ID if created successfully, None otherwise
    """
    existing_user = frappe.db.get_value("User", {"email": doc.personal_email}, "name")

    user = None
    if existing_user:
        user = frappe.get_doc("User", existing_user)
    else:
        user = frappe.get_doc(
            {
                "doctype": "User",
                "email": doc.personal_email,
                "first_name": doc.first_name or "Volunteer",
                "last_name": doc.last_name or "",
                "enabled": 1,
                "user_type": "System User",
                "send_welcome_email": 0,
                "default_app": "non_profit",
            }
        )
        user.insert(ignore_permissions=True)
    try:
        user.role_profile_name = "Volunteer"
        user.module_profile = "Volunteer"
        user.save(ignore_permissions=True)

        frappe.db.commit()
        return user.name

    except Exception as e:
        frappe.log_error(f"Error creating user for employee {doc.name}", f"{str(e)}")
        return None


def update_fields_from_applicant(doc: Document) -> bool:
    """
    Update employee fields from linked Job Applicant document
    Only updates if the employee field is empty
    Returns True if any fields were updated
    """
    try:
        job_applicant = frappe.get_doc("Job Applicant", doc.job_applicant)
        updated = False

        field_mapping = {
            "company": "company",
            "gender": "gender",
            "blood_group": "blood_group",
            "marital_status": "marital_status",
            "date_of_birth": "date_of_birth",
            "surname": "last_name",
            "other_names": "first_name",
            "email_id": "personal_email",
            "phone_number": "cell_number",
            "cover_letter": "bio",
            # "profile_photo": "image"
        }

        for applicant_field, employee_field in field_mapping.items():
            if (
                hasattr(job_applicant, applicant_field)
                and job_applicant.get(applicant_field)
                and not doc.get(employee_field)
            ):
                doc.set(employee_field, job_applicant.get(applicant_field))
                updated = True

        return updated

    except Exception as e:
        frappe.log_error(
            f"Error updating from job applicant {doc.job_applicant}", f"{str(e)}"
        )
        return False

File: non_profit/non_profit/overrides/client/employee.js
// Copyright (c) 2025, Frappe and contributors
// For license information, please see license.txt

frappe.ui.form.on("Employee", {
  refresh(frm) {
    if (frm.doc.job_applicant) {
      fetchJobApplicantDetails(frm);
    }

    if (!frm.doc.date_of_joining) {
      frm.set_value("date_of_joining", frappe.datetime.get_today());
    }

    frappe.db
      .get_value("Member", { email_id: frm.doc.personal_email }, "name")
      .then((r) => {
        if (!(r && r.message && r.message.name)) {
          frm.add_custom_button(__("Create Member"), () => {
            frm
              .call({
                method: "non_profit.non_profit.api.create_member",
                args: { name: frm.doc.name },
              })
              .then(() => {
                frappe.show_alert({
                  message: __("Member created successfully"),
                  indicator: "green",
                });
                frm.reload_doc();
              });
          });
        }
      });
  },

  job_applicant(frm) {
    if (!frm.doc.job_applicant) return;
    fetchJobApplicantDetails(frm);
  },
});

function fetchJobApplicantDetails(frm) {
  frappe.model.with_doc("Job Applicant", frm.doc.job_applicant, function () {
    const job_applicant = frappe.get_doc(
      "Job Applicant",
      frm.doc.job_applicant
    );
    const update_fields = {};

    const same_fields = [
      "company",
      "gender",
      "blood_group",
      "marital_status",
      "date_of_birth",
    ];

    const field_mapping = {
      surname: "last_name",
      other_names: "first_name",
      email_id: "personal_email",
      phone_number: "cell_number",
      cover_letter: "bio",
      // profile_photo: "image",
    };

    same_fields.forEach((field) => {
      if (job_applicant[field] && !frm.doc[field]) {
        update_fields[field] = job_applicant[field];
      }
    });

    Object.entries(field_mapping).forEach(
      ([applicant_field, employee_field]) => {
        if (job_applicant[applicant_field] && !frm.doc[employee_field]) {
          update_fields[employee_field] = job_applicant[applicant_field];
        }
      }
    );

    if (
      job_applicant.applicant_name &&
      !frm.doc.first_name &&
      !frm.doc.last_name
    ) {
      const parts = job_applicant.applicant_name.trim().split(" ");
      if (parts.length > 1) {
        update_fields.first_name = parts.slice(0, -1).join(" ");
        update_fields.last_name = parts.slice(-1).join(" ");
      } else {
        update_fields.first_name = parts[0];
      }
    }

    if (Object.keys(update_fields).length > 0) {
      frm.set_value(update_fields);
    }
  });
}
